### PR TITLE
[HTML5] Synchronous main, better persistence, handlers fixes, optional full screen.

### LINKED ISSUE #33020: Engine locked "touchend" event in Web Exported game
**Godot version:**
3.1.1 
**OS/device including version:**
Android 9 on Xiaomi Redmi Note 6 (any browser), Ubuntu Linux 18.04, Window 10 (Yandex Browser in mobile debug mode) 
**Issue description:**
When any html elements (e.g. ads popup) displays over GodotCanvas by call JavaScript.eval(), no touchend JS event passed to that elements. This behaviour blocks onclick JS events and the user can't  click on html elements. Removing touchend handler by Chromium DevTools (handler assigned to game's js file) allows to click the element.
On desktop (mouse events) all works correctly. On mobile browser in desktop mode is the same bug.
**Steps to reproduce:**
Create any Godot app (even empty screen). Add JavaScript snippet that dinamically displays, e.g. button, alerting on click (no difference, either snippet added in HTML shell or hardcoded to GDScript or else). Call this snippet with JavaScript.eval() from GDScript.
Export app to HTML5, open on web-server and try to click button. 
**Minimal reproduction project:**
This app creates empty scene with timer and after 3 seconds from start displays button.
[YaDummy.zip](https://github.com/godotengine/godot/files/3764767/YaDummy.zip)

### LINKED ISSUE #37205: Default HTML-5 export option forces scaling to fit viewport, even for non-resizable games
**Godot version:**
3.2
**OS/device including version:**
MacOS
**Issue description:**
Default HTML-5 export option forces scaling to fit viewport, even for non-resizable games
**Steps to reproduce:**
Use HTML-5 export for any project, compare the visible screen in Godot with the HTML-5 version when running.
**Minimal reproduction project:**
Any project with a pre-specified window width/height, smaller than the viewport it'll be run in.

### Quick-Fix:

Open up the Project.html of your export in a text-editor.
Scroll down to the line 
```js
function adjustCanvasDimensions() {
```
change the function from

```js
function adjustCanvasDimensions() {
	var scale = window.devicePixelRatio || 1;
	var width = window.innerWidth;
	var height = window.innerHeight;
	canvas.width = width * scale;
	canvas.height = height * scale;
	canvas.style.width = width + "px";
	canvas.style.height = height + "px";
	}
```

to

```js
function adjustCanvasDimensions() {
	canvas.width = innerWidth;
	canvas.height = innerHeight;
        }
```

### LINKED ISSUE #39643: Saving a file too frequently in HTML5 export will prevent the save file from being saved permanently.
**Godot version:**
3.2.1 stable

**Issue description:**
I was saving a file in a game (basically autosave) once a second and it seemed to not work on HTML5. The save wouldn't survive a tab refresh.

**Steps to reproduce:**
Try my web export here: https://armanlex.itch.io/save-and-load-test
1. Click and drag the icons around. Click Save, wait 5 seconds, refresh the tab and the game remembers the save. This works fine.

2. Click and drag the icons around. Click save, wait only one second then reload the tab. The save disappears. You can even save, quickly move an icon, click load (which means it reads the file off the storage location), it loads correctly! Then refresh and it still forgets the save.

3. Enable the autosave (atm it's once a second) and drag the icons around. You can move around the icons for minutes, but if you reload non of the saves after pressing autosave will survive. One way to make the autosave work is to change tab, wait at least 5 seconds then close the game and then open it.

4. Open the web export on two tabs. Move the icons on tab A and save. No matter how long you wait you'll never be able to load the tab A save from tab B. But if you refresh tab B then you will be able to load the tab A save.

I mostly tested this on Firefox, but Chrome seems to have the same problem.

I did a bunch of googling and I came across the terms SessionStorage and LocalStorage. It kinda seems like the saved files first go to SessionStorage (doesn't survive refresh and is tab specific) and then few seconds later to LocalStorage (does survive refresh, not tab specific). But that's just speculation from my part.

**Minimal reproduction project:**
Try my web export here: https://armanlex.itch.io/save-and-load-test
And here are the project files: https://github.com/Armanlex/save-and-load-test2

**Edit:**

I exported the project in 3.1.2 and recreated it for 3.0 and the same problem exist in those too.

### PR DESCRIPTION
**Pull request by Faless**
This PR has most of the things in #41097 (excluding the virtual keyboard support):

- Expose request_quit method to JS (sync with 3.2)
- Refactor event handlers, moving add/remove logic to native/utils.js
- window event handlers no longer useCapture (fixes #33020)
- Canvas resize option now available on export (fixes #37205).
- Add the override keyword to javascript platform

Plus, a rework of the initialization process to better handle file system synchronization (last commit, fixes #39643):

The engine now expects to emscripten FS to be setup and sync-ed before
main is called. This is exposed via `Module["initFS"]` which also allows
to setup multiple persistence paths (internal use only for now).

Additionally, FS syncing is done **once** for every loop if at least one
file in a persistent path was open for writing and closed, and if the FS
is not syncing already.

This should potentially fix issues reported by users where "autosave"
would not work on the web (never calling `syncfs` because of too many
writes).

There is a `3.2` branch of all these changes for testing and easy cherry picking: https://github.com/Faless/godot/tree/js/3.0_sync_fs_size_handlers

EDIT: ~Converted to draft, there is a bug when de-registering events. Will update soonish~ Fixed

### CODE REVIEW

**Comment by Faless** [comment]
Fixed silly bug in event listeners add/remove helpers.

File: platform/javascript/javascript_main.cpp [hunk 1]
```diff
@@ -32,13 +32,20 @@
...
 
 #include <emscripten/emscripten.h>
 #include <stdlib.h>
 
 static OS_JavaScript *os = nullptr;
 static uint64_t target_ticks = 0;
 
+extern "C" EMSCRIPTEN_KEEPALIVE void _request_quit_callback(char *p_filev[], int p_filec) {
+	if (os && os->get_main_loop()) {
+		os->get_main_loop()->notification(Node::NOTIFICATION_WM_CLOSE_REQUEST);
```
Comment (akien-mga) [inline platform/javascript/javascript_main.cpp:45]: There seems to be `DisplayServer::WINDOW_EVENT_CLOSE_REQUEST`, which is what triggers emitting `NOTIFICATION_WM_CLOSE_REQUEST` in `Windows`. Maybe this is better suited here than emitting a Node notification?

**Review by akien-mga — APPROVED** [review]
I left one tiny comment but otherwise this is good to merge.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**